enhancements/network/primary-user-defined-networks-for-virtualization-workloads.md

```diff
@@ -0,0 +1,568 @@
+validate the request - i.e. are the requested IPs in the configured ranges of
+the UDN network ? If not, the `IPAMClaim` condition will be set accordingly,
+and allocating the OVN annotation for the pod will fail. The network controller
+will keep attempting to reconcile the pod using exponential backoff until it
+possibly succeeds - in the case the user updates the
+`IPAMClaim.Spec.IPRequests` attribute.
+
+If the IPs being requested are in the configured subnet range of the network,
+the OVN-Kuberntes network controller proceeds to allocate the requested IP
+addresses; it might fail in case the requested IP addresses are already
+allocated in the network. If so, the pod will crashloop, and the `IPAMClaim`
+conditions updated to reflect this.
+
+In the case it succeeds, the `IPAMClaim` conditions are updated w/ a success
+condition, and its `IPAMCLaim.Status.IPs` updated accordingly (as happens today)
+for secondary networks.
+
+### Workflow Description
+
+#### Pre-requirements - configure the primary UDN for a namespace
+The user will need to perform the usual steps to configure a primary UDN for
+their namespace - that is, to provision a `UserDefinedNetwork` of layer2
+topology. While the user could use a `network-attachment-definition` CRD for
+this step, we are assuming (and recommending) the usage of the
+`UserDefinedNetwork` CRD - because we like statically typed things. You should
+too. You can find an example below with the minimal required data.
+
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: UserDefinedNetwork
+metadata:
+  name: safe-ground
+  namespace: tenantblue
+spec:
+  topology: Layer2
+  layer2:
+    role: Primary
+    subnets:
+    - 192.168.0.0/16
+    ipamLifecycle: Persistent
+```
+
+Keep in mind that the *user* mentioned above would **need** be an admin in case
+of a cluster-wide network - i.e. a network spanning multiple namespaces.
+
+#### Provisioning a VM object
+The user (VM owner) would then provision a VM in the cluster. You can find
+below an example of a VM:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-a
+  namespace: tenantblue
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: containerdisk
+          - disk:
+              bus: virtio
+            name: cloudinitdisk
+          interfaces:
+          - name: passtnet
+            binding:
+              name: passt
+          rng: {}
+        resources:
+          requests:
+            memory: 2048M
+      networks:
+      - name: passtnet
+        pod: {}
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - containerDisk:
+          image: quay.io/kubevirt/fedora-with-test-tooling-container-disk:v1.3.1
+        name: containerdisk
+      - cloudInitNoCloud:
+          networkData: |
+            version: 2
+            ethernets:
+              eth0:
+                dhcp4: true
```

Comment:
can be dropped right ?

Reply:
Thanks; good suggestions from both of you.